java-frontend/src/main/java/org/sonar/java/caching/ContentHashCache.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.caching;
+
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.batch.sensor.SensorContext;
+import org.sonar.api.batch.sensor.cache.ReadCache;
+import org.sonar.api.batch.sensor.cache.WriteCache;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+import java.io.IOException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+public class ContentHashCache {
+
+  private static final Logger LOG = Loggers.get(ContentHashCache.class);
+  private static final String CONTENT_HASH_KEY = String.format("java:contentHash:%s:", FileHashingUtils.HASH_ALGORITHM);
+  private static final String HASH_COMPUTE_FAIL_MSG = "Failed to compute content hash for file %s";
+
+  private ReadCache readCache;
+  private WriteCache writeCache;
+  private final boolean enabled;
+
+  public ContentHashCache(SensorContext context) {
+    CacheContextImpl cacheContext = CacheContextImpl.of(context);
+    enabled = cacheContext.isCacheEnabled();
+
+    if (enabled) {
+      readCache = context.previousCache();
+      writeCache = context.nextCache();
+    }
+  }
+
+  public boolean hasSameHashCached(InputFile inputFile) {
+    if (!enabled) {
+      LOG.trace("Cannot read from cache when disabled");
```

Comment:
Let's make the log message more specific here
```suggestion
      LOG.trace(() -> "Cannot read cached hashes when the cache is disabled (" + inputFile.key() + ").");
```